elkscmd/cron/crontab.1
.Dd October 16, 2007
.Os ELKS
.Dt CRONTAB 1
.Sh NAME
.Nm crontab
.Nd manage crontab files
.Sh SYNOPSIS
.Nm
.Op Fl u Ar user
.Op Pa file
.br
.Nm
.Op Fl u Ar user
.Bro
\& 
.Fl l ,
.Fl r ,
.Fl e
\& 
.Brc
.Sh DESCRIPTION
The
.Nm
utility is used to update, delete, or display the
files that
.Xr cron 8
uses to schedule jobs.  If invoked without options,
.Nm
copies the specified file, or the standard input
if no file is specified, into the directory that
holds all the system
.Xr crontabs 5 .
.Pp
The following options are supported:
.Bl -tag -width Fl XX
.It Fl e
Edits a copy of the current user's crontab file, or creates an
empty file to edit if their crontab does not exist.  When editing
is complete, the file is install as the user's crontab file.
.Pp The environment variables
.Em \&$EDITOR
and
.Em \&$VISUAL
determines which editor is used
.Po
.Em \&$EDITOR ,
then
.Em \&$VISUAL ,
and if neither of these is defined,
.Em vi
.Pc
.It Fl l
Lists the crontab file for the current user.
.It Fl r
Removes the crontab file for the current user.
.It Fl u Ar user
If you are logged in as the superuser,  you can
use 
.Fl u
to specify a particular user's crontab to update,
delete, or list.
.El
.Sh BUGS
If you inadvertently enter the crontab command with no
arguments, you will lose your existing crontab unless
you exit with a
.Sy Control-C .
.Nm Crontab
attempts to catch this bug by refusing to overwrite a
crontab with an empty one, but it's better to be safe
than sorry.
.Sh FILES
.Bl -tag -width "/var/spool/cron"
.It Pa "/var/spool/cron"
Where 
.Xr cron 8
looks for crontabs.
.El
.Sh SEE ALSO
.Xr crontab 5 ,
.Xr cron 8 .
.Sh AUTHOR
David Parsons (orc@pell.portland.or.us)


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/cron/crontab.c
/*
 * crontab: read/write/delete/edit crontabs
 */
#include "config.h"

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <pwd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <utime.h>
#include <ctype.h>
#if HAVE_LIBGEN_H
#   include <libgen.h>
#endif

#include "cron.h"

#ifndef seteuid
    #define seteuid setuid
#endif

char *pgm = "crontab";
FILE *in_tmp;
#define CRONTMP "crontmp.tmp"
extern int interactive;

void
usage()
{
    fprintf(stderr, "usage: %s [-u user] -{l,r,e}\n"
                    "       %s [-u user] file\n", pgm, pgm);
    exit(1);
}


/* pick up our superpowers (if any)
 */
void
superpowers()
{
    if (getuid() != geteuid()) {
	setuid(geteuid());
	setgid(getegid());
    }
}


/* become super, then cat a file out of CRONDIR
 */
int
cat(char *file)
{
    register int c;
    FILE *f;

    superpowers();
    xchdir(CRONDIR);

    if ( f = fopen(file, "r") ) {
	while ( (c = fgetc(f)) != EOF ) {
	    putchar(c);
	}
	fclose(f);
	return 1;
    }
    return 0;
}


/* put a new crontab into CRONDIR
 */
int
newcrontab(char *file, char *name)
{
    static crontab tab = { 0 };
    int tempfile = 0;
    register int c;
    FILE *out;
    struct utimbuf touch;
    int save_euid = geteuid();

    umask(077);

    if ( strcmp(file, "-") == 0 ) {
	long size = 0;
	////if ( (in_tmp = tmpfile()) == 0 ) {
        if ( (in_tmp = fopen(CRONTMP, "wb+")) == 0 ) {
	    error("can't create tempfile: %s", strerror(errno));
	    return 0;
	}
	while ( ((c = getchar()) != EOF) && !ferror(in_tmp) ) {
	    ++size;
	    fputc(c, in_tmp);
	}
	if (ferror(in_tmp) || (fflush(in_tmp) == EOF) || fseek(in_tmp, 0L, SEEK_SET) == -1 ) {
	    fclose(in_tmp);
	    error("can't store crontab in tempfile: %s", strerror(errno));
	    return 0;
	}
	if (size == 0) {
	    fatal("no input; use -r to delete a crontab");
	}
	tempfile = 1;
    }
    else {
	seteuid(getuid());
	in_tmp = fopen(file, "r");
	seteuid(save_euid);
	if ( in_tmp == 0 ) {
	    error("can't open %s: %s", file, strerror(errno));
	    return 0;
	}
    }

    zerocrontab(&tab);
    if ( !readcrontab(&tab, in_tmp) ) {
	fclose(in_tmp);
	return 0;
    }
    if ( tempfile && (tab.nrl == 0) && (tab.nre == 0) ) {
	error("no jobs or environment variables defined; use -r to delete a crontab");
	return 0;
    }
    if ( fseek(in_tmp, 0L, SEEK_SET) == -1 ) {
	fclose(in_tmp);
	error("file error: %s", strerror(errno));
	return 0;
    }

    superpowers();
    xchdir(CRONDIR);

    if ( out = fopen(name, "w") ) {
	while ( ((c = fgetc(in_tmp)) != EOF) && (fputc(c,out) != EOF) )
	    ;
	if ( ferror(out) || (fclose(out) == EOF) ) {
	    unlink(name);
	    fatal("can't write to crontab: %s", strerror(errno));
	}
	fclose(in_tmp);
	time(&touch.modtime);
	time(&touch.actime);
	utime(".", &touch);
	exit(0);
    }
    fatal("can't write to crontab: %s", strerror(errno));
    return 0;
}


static char vitemp[24];

/* erase the tempfile when the program exits
 */
static void
nomorevitemp()
{
    unlink(vitemp);
    fclose(in_tmp);
    unlink(CRONTMP);
}


/* Use a visual editor ($EDITOR,$VISUAL, or vi) to edit a crontab,
 * then copy the new one into CRONDIR
 */
void
visual(struct passwd *pwd)
{
    char ans[20];
    char commandline[80];
    char *editor;
    char *yn;
    struct stat st;
    int save_euid = geteuid();

    xchdir(pwd->pw_dir ? pwd->pw_dir : "/tmp");
    strcpy(vitemp, "/tmp/.crontab.XXXXXX");
    seteuid(getuid());
    mktemp(vitemp);

    atexit(nomorevitemp);

    if ( ((editor = getenv("EDITOR")) == 0) && ((editor = getenv("VISUAL")) == 0) )
	editor="vi";

    if ( access(editor, X_OK) == 0 )
	fatal("%s: %s", editor, strerror(errno));

    sprintf(commandline, "crontab -l > %s", vitemp);
    system(commandline);

    if ( (stat(vitemp, &st) != -1) && (st.st_size == 0) )
	unlink(vitemp);

    sprintf(commandline, "%s %s", editor, vitemp);
    while (1) {
	if ( system(commandline) == -1 )
	    fatal("running %s: %s", editor, strerror(errno));
	seteuid(save_euid);
	if ( (stat(vitemp, &st) == -1) || (st.st_size == 0)
	                                 || newcrontab(vitemp, pwd->pw_name))
	    exit(0);
	seteuid(getuid());
	do {
	    fprintf(stderr, "Do you want to retry the same edit? ");
	    fgets(ans, sizeof ans, stdin);
	    yn = firstnonblank(ans);
	    *yn = toupper(*yn);
	    if (*yn == 'N')
		exit(1);
	    if (*yn != 'Y')
		fprintf(stderr, "(Y)es or (N)o; ");
	} while (*yn != 'Y');
    }
}


/* crontab
 */
int
main(int argc, char **argv)
{
    int opt;
    int edit=0,remove=0,list=0;
    char *user = 0;
    char whoami[20];
    char *what = "update";
    struct passwd *pwd;

    opterr = 1;

#if HAVE_BASENAME
    pgm = basename(argv[0]);
#else
    if (pgm = strrchr(argv[0], '/') )
	++pgm;
    else
	pgm = argv[0];
#endif

    if ( (pwd = getpwuid(getuid())) == 0 )
	fatal("you don't seem to have a password entry?\n");
    strncpy(whoami, pwd->pw_name, sizeof whoami);

    while ( (opt=getopt(argc,argv, "elru:")) != EOF ) {
	switch (opt) {
	case 'e': edit = 1; what = "edit"; break;
	case 'l': list = 1; what = "list"; break;
	case 'r': remove = 1; what = "remove"; break;
	case 'u': user = optarg; break;
	default:  usage();
	}
    }
    if ( edit + list + remove > 1 )
	usage();

    if (user) {
	if ( (pwd = getpwnam(user)) == 0 )
	    fatal("user %s does not have a password entry.\n", user);
    }
    else if ( (pwd = getpwuid(getuid())) == 0 )
	fatal("you don't seem to have a password entry?\n");

    if ( (getuid() != 0) && (pwd->pw_uid != getuid()) )
	fatal("you may not %s %s's crontab.\n", what, pwd->pw_name);

    argc -= optind;
    argv += optind;

    interactive=0;
    info((strcmp(whoami,pwd->pw_name) == 0) ? "(%s) %s" : "(%s) %s (%s)",
				whoami, what, pwd->pw_name);
    interactive=1;
    if (list) {
	if ( !cat(pwd->pw_name) )
	    fatal("no crontab for %s", pwd->pw_name);
    }
    else if (remove) {
	superpowers();
	xchdir(CRONDIR);
	if (unlink(pwd->pw_name) == -1)
	    fatal("%s: %s", pwd->pw_name, strerror(errno));
    }
    else if (edit)
	visual(pwd);
    else if ( !newcrontab( argc ? argv[0] : "-", pwd->pw_name) )
	exit(1);
    exit(0);
}
